Mathlib/ModelTheory/Substructures.lean

```diff
@@ -885,8 +887,292 @@ theorem range_subtype (S : L.Substructure M) : S.subtype.toHom.range = S := by
+noncomputable def dom_cod_top_toEquiv {f : M ≃ₚ[L] N} (h_dom : f.source = ⊤)
+    (h_cod : f.target = ⊤) : M ≃[L] N :=
+  (topEquiv (M := N)).comp ((h_dom ▸ h_cod ▸ f.equiv).comp (topEquiv (M := M)).symm)
+
+@[simp]
+theorem dom_cod_top_toEquiv_toEmbedding {f : M ≃ₚ[L] N} (h_dom : f.source = ⊤)
+    (h_cod : f.target = ⊤) :
+    (dom_cod_top_toEquiv h_dom h_cod).toEmbedding = dom_top_toEmbedding h_dom := by
+  rcases f with ⟨dom, cod, g⟩
+  cases h_dom
+  cases h_cod
+  rfl
+
+/--Map of a self-partial equivalence through an embedding.-/
+noncomputable def map (f : M ↪[L] N) (g : M ≃ₚ[L] M) : N ≃ₚ[L] N where
+  source := g.source.map f.toHom
+  target := g.target.map f.toHom
+  equiv := (f.substructureEquivMap g.target).comp <|
+    g.equiv.comp (f.substructureEquivMap g.source).symm
+
+@[simp]
+theorem map_dom (f : M ↪[L] N) (g : M ≃ₚ[L] M) : (g.map f).source = g.source.map f.toHom := rfl
+
+@[simp]
+theorem map_cod (f : M ↪[L] N) (g : M ≃ₚ[L] M) : (g.map f).target = g.target.map f.toHom := rfl
+
+theorem map_commutes (f : M ↪[L] N) (g : M ≃ₚ[L] M) :
+    (g.map f).equiv.comp (f.substructureEquivMap g.source) =
+      (f.substructureEquivMap g.target).comp g.equiv := by
+  unfold map
+  ext
+  simp only [Equiv.comp_apply, Equiv.symm_apply_apply, Embedding.substructureEquivMap_apply]
+
+theorem map_commutes_apply (f : M ↪[L] N) (g : M ≃ₚ[L] M) (m : g.source) :
+    (g.map f).equiv ⟨f m, g.source.apply_coe_mem_map _ _⟩ =
+      ⟨f (g.equiv m), g.target.apply_coe_mem_map _ _⟩ := by
+  exact congr_fun (congr_arg DFunLike.coe (g.map_commutes f)) m
+
+theorem map_monotone (f : M ↪[L] N) : Monotone (fun g : M ≃ₚ[L] M ↦ g.map f) := by
+  intro g g' h
+  rw [le_iff]
+  use Substructure.monotone_map (le_dom h)
+  use Substructure.monotone_map (le_cod h)
+  rintro ⟨x, hx⟩
+  unfold map
+  let ⟨u, u_mem, eq_u_x⟩ := mem_map.2 hx
+  cases eq_u_x
+  apply Subtype.coe_injective
+  simp only [Embedding.coe_toHom, Equiv.comp_apply, coe_inclusion, map_coe, Set.coe_inclusion,
+    Embedding.substructureEquivMap_apply, Set.inclusion_mk, EmbeddingLike.apply_eq_iff_eq]
+  let ⟨_, _, eq⟩ := le_iff.1 h
+  have eq := congr_arg (Subtype.val) (eq ((Equiv.symm (Embedding.substructureEquivMap f g.source))
+    { val := f u, property := (g.source.mem_map).2 ⟨u, u_mem, rfl⟩}))
+  simp only [coe_inclusion, Set.coe_inclusion] at eq
+  rw [← coe_inclusion] at eq
+  rw [eq, Subtype.coe_inj]
+  apply congr_arg g'.equiv
+  apply Subtype.coe_injective
+  change subtype _ ((Equiv.symm (Embedding.substructureEquivMap f g.source))
+    (f.substructureEquivMap g.source ⟨u, u_mem⟩)) =
+    subtype _ ((Equiv.symm (Embedding.substructureEquivMap f g'.source))
+      (f.substructureEquivMap g'.source ⟨u, le_dom h u_mem⟩))
+  simp only [Equiv.symm_apply_apply, coeSubtype]
+
+end PartialEquiv
+
 end Substructure
 
+namespace Embedding
+
+/--Given an embedding, returns the corresponding partial equivalence with `⊤` as domain.-/
+noncomputable def toPartialEquiv (f : M ↪[L] N) : M ≃ₚ[L] N :=
```

Comment:
Ah! Your `PartialEquiv.map` is like `Embedding.toPartialEquiv f \comp g \comp (Embedding.toPartialEquiv f).symm`.

Do you think you could *define* it like that? /maybe not define it at all

Reply:
At the moment, there's no notion of composition at all for `PartialEquiv` - in general, it should give you a partial equiv whose domains are intersections of the domains of things being composed, and that might get slightly slippery.
It does make sense what it should be for composing on the left or the right with an embedding. If one of those is defined, the other can be defined using `symm` a few times.

In any event, I'm going to go poke around with the PRs depending on this one to see how much they use this specific formulation of `map`.

Reply:
Yeah, I would potentially suggest removing `map` from this PR and move to a later PR which needs it. Then we'll have a better idea of what exactly we want from it.

Reply:
It looks like it's only needed for preliminary work that was done on showing existence of Fraïssé limits, and not for uniqueness, so I'm happy to trim it for now to get the PR through.